### PR TITLE
soft kernel return code and crash detection in zocl scheduler

### PR DESCRIPTION
**Pull request by sarab96**
soft kernel return code and crash detection in zocl scheduler
Rest of changes for host side will be in separate PR

### CODE REVIEW

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 1]
```diff
@@ -60,6 +60,13 @@ static int    sched_loop_cnt;
 static struct scheduler g_sched0;
 static struct sched_ops penguin_ops;
 static struct sched_ops ps_ert_ops;
+#define SK_CRASHED -1
+#define SK_ERROR -2
+#define SK_DONE 1
+#define SK_RUNNING 2
+#define AP_DONE 2
+#define AP_START 1
+
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c:69]: AP_DONE and AP_START is defined in zocl_cu.h

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 2]
```diff
@@ -60,6 +60,13 @@ static int    sched_loop_cnt;
 static struct scheduler g_sched0;
 static struct sched_ops penguin_ops;
 static struct sched_ops ps_ert_ops;
+#define SK_CRASHED -1
+#define SK_ERROR -2
+#define SK_DONE 1
+#define SK_RUNNING 2
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c:66]: Can we define SK related stuff in zocl_sk.h?

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 3]
```diff
@@ -1454,40 +1461,61 @@ cu_done(struct sched_cmd *cmd)
...
+	struct soft_krnl *sk = zdev->soft_kernel;
+	struct soft_cu *scu;
+	uint32_t *cu_regfile;
+	struct ert_start_kernel_cmd *skc;
+
+	skc = (struct ert_start_kernel_cmd *)cmd->packet;
+	scu = sk->sk_cu[cmd->cu_idx];
+	cu_regfile = scu->sc_vregs;
+
+        skc->data[skc->extra_cu_masks + 1] = cu_regfile[1];
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c:1478]: Also, I suggest we add some comment and document it later that the CU (include soft CU and hard CU) return code is put into a fixed place of ert_start_kernel_cmd structure. 
There seems to be an issue for the place at pkt->data[pkt->extra_cu_maskes + 1] when pkt->count = (0 or 1). It overlaps with the command state. see ert_start_kernel_timestamps

**Review by larry9523 — CHANGES_REQUESTED** [review]
1. Please fix the indent (there are quite a few places have indent issue)
2. Please use /* */ instead of // in kernel code if you can
3. Please consider this return code to be generic across hard cu and soft cu

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 4]
```diff
@@ -1454,40 +1461,61 @@ cu_done(struct sched_cmd *cmd)
...
+}
+
+inline int32_t
```
[the remark this answers is not in the document]
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c:1474]: I doubt our driver will work with 8 bit or 16 bit systems. In any case ERT packet entries & AXI register all require 32 bits so better to use 32 bits for these areas.

File: src/runtime_src/core/include/ert.h [hunk 5]
```diff
@@ -364,6 +364,8 @@ enum ert_cmd_state {
   ERT_CMD_STATE_SUBMITTED = 7,
   ERT_CMD_STATE_TIMEOUT = 8,
   ERT_CMD_STATE_NORESPONSE = 9,
+  ERT_CMD_STATE_SKERROR = 10, //Check for error return code from Soft Kernel
+  ERT_CMD_STATE_SKCRASHED = 11, //Soft kernel has crashed
```
[the remark this answers is not in the document]
Comment (sarab96) [reply src/runtime_src/core/include/ert.h:368]: Crashed soft kernel and error return code from soft kernel are unique so need these. Did not reuse .._ERROR as that is for error during submission to ERT. And better to differentiate crashed from other error codes or error conditions.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 1 above)
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: done

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 2 above)
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: done

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: done

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 6]
```diff
@@ -1454,40 +1461,61 @@ cu_done(struct sched_cmd *cmd)
...
 scu_done(struct sched_cmd *cmd)
 {
 	struct drm_zocl_dev *zdev = cmd->ddev->dev_private;
 	int cu_idx = cmd->cu_idx;
 	struct soft_krnl *sk = zdev->soft_kernel;
 	u32 *virt_addr;
+        int32_t ret = SK_ERROR;
```
[the remark this answers is not in the document]
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: done

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 7]
```diff
@@ -2722,8 +2751,24 @@ ps_ert_query(struct sched_cmd *cmd)
...
+                        break;
+                    case SK_DONE:
 			mark_cmd_complete(cmd, ERT_CMD_STATE_COMPLETED);
+                        break;
+                    case SK_RUNNING:
+                        break;
+                    default:
+		        mark_cmd_complete(cmd, ERT_CMD_STATE_ERROR);
+		        DRM_ERROR("unknown opcode %d", opc);
+                        break;
```
[the remark this answers is not in the document]
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: done

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 3 above)
Comment (sarab96) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: updated. Please check again

**Review by larry9523 — APPROVED** [review]

**Pull request merged**